src/hypothesis/strategies.py

```diff
@@ -857,18 +857,21 @@ def recursive(base, extend, max_leaves=100):
 def permutations(values):
     """Return a strategy which returns permutations of the collection
     "values"."""
+    from hypothesis.internal.conjecture.utils import integer_range
+
     values = list(values)
     if not values:
-        return just(()).map(lambda _: [])
-
-    def build_permutation(swaps):
-        initial = list(values)
-        for i, j in swaps:
-            initial[i], initial[j] = initial[j], initial[i]
-        return initial
-    n = len(values)
-    index = integers(0, n - 1)
-    return lists(tuples(index, index), max_size=n ** 2).map(build_permutation)
+        return builds(list)
+
+    class PermutationStrategy(SearchStrategy):
+
+        def do_draw(self, data):
+            result = list(values)
+            for i in range(len(result)):
+                j = integer_range(data, i, len(result) - 1)
```

Comment:
Wait, it would help if I read the title of the pull request, and thought about how Fisher-Yates permutations work.

Worth a comment to explain the basis of the shuffling algorithm? (Just a namedrop would be enough.)